From 70ee4d487293cf3a5662bcc01ead5ea32eb93681 Mon Sep 17 00:00:00 2001
From: ksooo <kai.sommerfeld@gmx.com>
Date: Fri, 23 Oct 2026 10:07:09 +0200
Subject: [PATCH] [PVR] Channels OSD: Fix channel up/down after group switch
 w/o channel switch

Switching the group in the channels OSD makes the new group the active
group. If the OSD was closed without selecting a channel from that group,
channel up/down could not find the playing channel in the active group
and did nothing.

Restore the playing channel's group on close in this case.

Fixes #29932

(cherry picked from commit 013fadb1fd35e208b30da6736292efba7de1bf08)
---
 xbmc/pvr/PVRPlaybackState.h                  | 12 ++++++------
 xbmc/pvr/dialogs/GUIDialogPVRChannelsOSD.cpp | 12 ++++++++++--
 2 files changed, 16 insertions(+), 8 deletions(-)

diff --git a/xbmc/pvr/PVRPlaybackState.h b/xbmc/pvr/PVRPlaybackState.h
index 39f90fc..0fe8152 100644
--- a/xbmc/pvr/PVRPlaybackState.h
+++ b/xbmc/pvr/PVRPlaybackState.h
@@ -222,6 +222,12 @@ public:
    */
   void SetActiveChannelGroup(const std::shared_ptr<CPVRChannelGroup>& group);
 
+  /*!
+   * @brief Set the active group to the group of the supplied channel group member.
+   * @param channel The channel group member
+   */
+  void SetActiveChannelGroup(const std::shared_ptr<CPVRChannelGroupMember>& channel);
+
   /*!
    * @brief Get the active channel group.
    * @param bRadio True to get the active radio group, false to get the active TV group.
@@ -271,12 +277,6 @@ private:
    */
   std::unique_ptr<CFileItem> GetNextAutoplayItem(const CFileItem& item);
 
-  /*!
-   * @brief Set the active group to the group of the supplied channel group member.
-   * @param channel The channel group member
-   */
-  void SetActiveChannelGroup(const std::shared_ptr<CPVRChannelGroupMember>& channel);
-
   /*!
    * @brief Updates the last watched timestamps of the channel and group which are currently playing.
    * @param channel The channel which is updated
diff --git a/xbmc/pvr/dialogs/GUIDialogPVRChannelsOSD.cpp b/xbmc/pvr/dialogs/GUIDialogPVRChannelsOSD.cpp
index d63c3a7..e1122ab 100644
--- a/xbmc/pvr/dialogs/GUIDialogPVRChannelsOSD.cpp
+++ b/xbmc/pvr/dialogs/GUIDialogPVRChannelsOSD.cpp
@@ -100,8 +100,16 @@ void CGUIDialogPVRChannelsOSD::OnDeinitWindow(int nextWindowID)
 {
   if (m_group)
   {
-    CServiceBroker::GetPVRManager().Get<PVR::GUI::Channels>().SetSelectedChannelPath(
-        m_group->IsRadio(), m_viewControl.GetSelectedItemPath());
+    CPVRManager& pvrMgr = CServiceBroker::GetPVRManager();
+    const std::shared_ptr<CPVRChannelGroupMember> playingMember =
+        pvrMgr.PlaybackState()->GetPlayingChannelGroupMember();
+
+    // Group was switched without selecting one of its channels.
+    if (playingMember && playingMember->GroupID() != m_group->GroupID())
+      pvrMgr.PlaybackState()->SetActiveChannelGroup(playingMember);
+    else
+      pvrMgr.Get<PVR::GUI::Channels>().SetSelectedChannelPath(m_group->IsRadio(),
+                                                              m_viewControl.GetSelectedItemPath());
 
     // next OnInitWindow will set the group which is then selected
     m_group.reset();
-- 
2.53.0

